### PR TITLE
Minor improvement of snapshots handling

### LINKED ISSUE #1082: set_snapshots reindexing to multiindex resets all data
### Version Checks (indicate both or one)

- [X] I have confirmed this bug exists on the lastest [release](https://github.com/pypsa/pypsa/releases) of PyPSA.

- [X] I have confirmed this bug exists on the current [`master`](https://github.com/pypsa/pypsa/tree/master) branch of PyPSA.

### Issue Description

If I want to reindex a network to add investment periods as a multiindex, all the time varying data in the network get's reset to defaults even if the inner level of the new multiindex identically matches the original index.

### Reproducible Example

```python
import pandas as pd
import pypsa

n = pypsa.examples.ac_dc_meshed(from_master=True)

periods = n.snapshots.year.unique()
timesteps = n.snapshots
n.set_snapshots(pd.MultiIndex.from_product((periods, timesteps)))

n.generators_t.p_max_pu  # now all 1.0
```

### Expected Behavior

Time varying network data should be unaffected when reindexing with a multiindex where the inner level of the new multiindex matches the original index.

I _think_ this is a relatively simple fix to add `level=-1` in `dynamic[k].reindex` call. 

https://github.com/PyPSA/PyPSA/blob/c69054626978d05b436c0978189c51f9e8b8ad73/pypsa/components.py#L727-L732

### Installed Versions

<details>

pypsa==0.31.1
pandas==2.2.3

</details>

### PR DESCRIPTION
**Pull request by lkstrp**
Closes #1082

## Changes proposed in this Pull Request
**Current state**
- Snapshots can be single or multi-indexed (have investment periods). This is currently represented by ambiguous index names:
  - Without investment periods: 
    - `n.snapshots.index.name` - `'snapshot'`
    - `n.snapshots.index.names`: `['snapshot']`
  - With investment periods:
    - `n.snapshots.index.name` - `'snapshot'`.
    - `n.snapshots.index.names` - `['period', 'timestep']`.
- The same timestep/snapshot data is referred to by both `timestep` and `snapshot`. This can lead to all kinds of bugs, especially with the upcoming `scenario` dimension, and should be harmonised.

**Two ways to fix this**
- Rename `timestep` to `snapshot`
  - Pro: Only renames the index for networks with multiple investment periods
  - Con: Still no unambiguous naming. Since `n.snapshots` is both `snapshot` and `period` data, it is an arbitray access to the timestep data. But renaming `n.snapshots` is very intrusive.
- Rename `snapshot` to `timestep`
  - Pro: Cleaner. Ambiguous `n.snapshots` would be resolved.  
  - Con: Would change the index name for all networks without multiple investment periods. I do not know how much this would break.

This PR implements the first option. But I would rather do the second and get rid of `snapshot` as an index name altogether.

**Multiindex names**
In any case, I would remove the single index name for multi-indexed snapshots. Since `idx.names==[idx.name]` for a single dimension, we can drop it for multi-indexed snapshots.  This would result in clear and unique dimension names for each dimension and fix the current behaviour.

Some tests fail because `linopy` can't handle a tuple of coordinates, both single indexed and multi indexed at the moment. This could be fixed here or in `linopy`. But let me know what you think first.

## Checklist

- [x] Unit tests for new features were added (if applicable).
- [ ] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

File: pypsa/utils.py [hunk 1]
```diff
@@ -11,12 +11,7 @@
     from pypsa import Network
 
 
-def as_index(
-    n: Network,
-    values: Any,
-    network_attribute: str,
-    index_name: str | None,
-) -> pd.Index:
+def as_index(n: Network, values: Any, network_attribute: str) -> pd.Index:
```
Comment (fneum) [inline pypsa/utils.py:14]: `index_name` should be removed from docstring

File: pypsa/utils.py [hunk 2]
```diff
@@ -60,7 +61,7 @@ def as_index(
 def deprecated_kwargs(**aliases: str) -> Callable:
     """
     Decorator for deprecated function and method arguments.
-    Based on solution from [here](https://stackoverflow.com/questions/49802412).
+    Based on solution fr [here](https://stackoverflow.com/questions/49802412).
```
Comment (fneum) [inline pypsa/utils.py:64]: not intended?

File: test/test_components.py
Comment (fneum) [inline test/test_components.py:1]: Mostly restructuring?

**Review by fneum — APPROVED** [review]
> This PR implements the first option. But I would rather do the second and get rid of snapshot as an index name altogether.

I agree with your decision to implement the first option for now (less intrusive). Perhaps with https://github.com/PyPSA/PyPSA/pull/1037 we can then rename the index level to `timestep`. The main reason why we used the term "snapshot" in the first place rather than "timestep" is because it could also be used for stochastic problem definitions. With #1037 the intended use is different.

**Comment by lkstrp** [comment]
> I agree with your decision to implement the first option for now (less intrusive). Perhaps with https://github.com/PyPSA/PyPSA/pull/1037 we can then rename the index level to timestep.

@fneum, hmm, not sure I agree. If people are using index names in n.snapshots, this is a breaking change and can't be made backwards compatible. Probably not a problem for many users, but still. Merging this pr with option 1, and fixing it later with 1037, would just mean breaking things twice for those users.
So I would rather go with option 2 here and put it already in `v0.32`.

**Comment by lkstrp** [comment]
For the history books:
This PR confuses dimension with dimension levels.
Periods and timesteps are just levels of the snapshots dimension, and this cannot be changed:
- `timesteps` are not unique for each investment_period, but contain year data. The year cannot be removed either, since timestamps cannot exist without year information.
- It is also possible to define different numbers of timesteps per investment period.
For them to be independent dimensions, both would be necessary.

A follow up PR will clarify that.

**Pull request merged**